README.md
# humanize-duration

Fork of [EvanHahn/HumanizeDuration.js](https://github.com/EvanHahn/HumanizeDuration.js). Convert millisecond durations to English and many other languages.

## Changes from upstream

* Add `zh` and `ru` language support
* Add TypeScript declaration file
* Remove languages we don't need

## Install

```
npm install @sector-labs/humanize-duration
```

## Usage

```js
var humanizeDuration = require('@sector-labs/humanize-duration')
humanizeDuration(12000) // '12 seconds'
```

### Options

| Option | Description |
|--------|-------------|
| `language` | Language for unit display (ISO 639-1 code) |
| `delimiter` | String to display between units (default: `', '`) |
| `spacer` | String to display between value and unit (default: `' '`) |
| `largest` | Maximum number of units to display |
| `units` | Array of units to use: `['y', 'mo', 'w', 'd', 'h', 'm', 's', 'ms']` |
| `round` | Round the smallest unit (default: `false`) |
| `decimal` | String to substitute for the decimal point |
| `conjunction` | String to include before the final unit |
| `serialComma` | Use serial comma with conjunction (default: `true`) |
| `unitMeasures` | Customize the value used to calculate each unit of time |

### Humanizers

Create a humanizer with default options:

```js
var spanishHumanizer = humanizeDuration.humanizer({
  language: 'es',
  units: ['y', 'mo', 'd']
})

spanishHumanizer(71177400000)  // '2 años, 3 meses, 2 días'
```

## License

Unlicense


HISTORY.md
3.15.1-sl.1 / 2018-06-25
========================

* new: Removed everything besides en, ur and ar


3.15.0 / 2018-06-12
===================

* new: Croatian support
* new: Lao support

3.14.0 / 2018-03-23
===================

* new: Urdu support

3.13.0 / 2018-03-19
===================

* new: Slovak support

3.12.1 / 2018-01-03
===================

* fix: Ukranian translation for weeks was incorrect

3.12.0 / 2017-11-07
===================

* new: Bulgarian support

3.11.0 / 2017-10-25
===================

* new: Persian/Farsi support

3.10.1 / 2017-07-21
===================

* fix: use singular form from -2 to 2 in French

3.10.0 / 2016-11-29
===================

* new: Icelandic support

3.9.1 / 2016-07-24
==================

* fix: Russian and Ukranian fixes

3.9.0 / 2016-06-10
==================

* new: Indonesian support
* new: Malay support

3.8.0 / 2016-05-19
==================

* new: `conjunctions` and `serialComma` options

* update: improve documentation

3.7.1 / 2016-04-26
==================

* fix: rounding had some more errors

3.7.0 / 2016-03-18
==================

* new: Vietnamese support

3.6.1 / 2016-02-19
==================

* update: readme should use single quotes

* fix: rounding had some errors (for example, you could get "1 day, 24 hours")
* fix: readme example was missing some quotes in its result

3.6.0 / 2016-02-12
==================

* new: Lithuanian support

* fix: add trailing semicolon to improve concatenation with other libraries

3.5.0 / 2016-01-13
==================

* new: Finnish support
* new: recommend [millisec](https://github.com/sungwoncho/millisec) module in readme

3.4.0 / 2015-12-11
==================

* new: Czech support

* update: minor performance improvements
* update: add "related modules" section to readme

* fix: add Greek to readme

3.3.0 / 2015-08-05
==================

* new: `unitMeasures` option

3.2.1 / 2015-07-22
==================

* fix: document `largest` option

3.2.0 / 2015-07-15
==================

* new: Ukrainian support

* fix: things would break in global strict mode

3.1.0 / 2015-06-25
==================

* new: Greek support

3.0.0 / 2015-05-29
==================

* new: `largest` option to get largest units

* update: languages can change decimal point (which can be overridden)
* update: fix some unintuitive `round` behavior
* update: specify units in shorthand rather than longhand
* update: use the Unlicense
* update: `nob` is now `no`
* update: use underscores instead of dashes for Chinese

* remove: half units

2.8.0 / 2015-05-10
==================

* new: `getSupportedLanguages`

2.7.0 / 2015-05-08
==================

* new: changelog
* new: Arabic support
* new: Italian support

* remove: (spoken) languages from package.json keywords


test/error.js
var humanizeDuration = require('..')
var assert = require('assert')

describe('error handling', function () {
  it('throws an error when passed a bad language in the function', function () {
    function humanizingWith (options) {
      return function () {
        humanizeDuration(10000, options)
      }
    }

    assert.throws(humanizingWith({ language: 'EN' }), Error)
    assert.throws(humanizingWith({ language: 'bad language' }), Error)
    assert.throws(humanizingWith({ language: '' }), Error)
    assert.throws(humanizingWith({ language: null }), Error)
  })

  it('throws an error when passed a bad language in a humanizer', function () {
    var h = humanizeDuration.humanizer({
      language: 'bad language'
    })

    function humanizing (options) {
      return function () {
        h(10000, options)
      }
    }

    assert.throws(humanizing(), Error)
    assert.throws(humanizing({ language: 'EN' }), Error)
    assert.throws(humanizing({ language: 'bad language' }), Error)
    assert.throws(humanizing({ language: '' }), Error)
    assert.throws(humanizing({ language: null }), Error)
    assert.doesNotThrow(humanizing({ language: 'ar' }), Error)
  })
})


test/humanizer.js
var humanizeDuration = require('..')
var humanizer = humanizeDuration.humanizer
var assert = require('assert')
var ms = require('ms')

describe('humanizer', function () {
  it('humanizes English when passed no arguments', function () {
    var h = humanizer()

    assert.equal(h(1000), '1 second')
  })

  it('humanizes English when passed an empty object', function () {
    var h = humanizer({})

    assert.equal(h(1000), '1 second')
  })

  it('can change the delimiter', function () {
    var h = humanizer({ delimiter: '+' })

    assert.equal(h(0), '0 seconds')
    assert.equal(h(1000), '1 second')
    assert.equal(h(363000), '6 minutes+3 seconds')
  })

  it('can change the spacer', function () {
    var h = humanizer({ spacer: ' whole ' })

    assert.equal(h(0), '0 whole seconds')
    assert.equal(h(1000), '1 whole second')
    assert.equal(h(260040000), '3 whole days, 14 whole minutes')
  })

  it('can use a conjunction', function () {
    var h = humanizer({ conjunction: ' and ' })

    assert.equal(h(0), '0 seconds')
    assert.equal(h(1000), '1 second')
    assert.equal(h(260040000), '3 days and 14 minutes')
    assert.equal(h(10874000), '3 hours, 1 minute, and 14 seconds')
  })

  it('can use a conjunction without a serial comma', function () {
    var h = humanizer({
      conjunction: ' & ',
      serialComma: false
    })

    assert.equal(h(1000), '1 second')
    assert.equal(h(260040000), '3 days & 14 minutes')
    assert.equal(h(10874000), '3 hours, 1 minute & 14 seconds')
  })

  it('can change the units', function () {
    var h = humanizer({ units: ['d'] })

    assert.equal(h(0), '0 days')
    assert.equal(h(ms('6h')), '0.25 days')
    assert.equal(h(ms('7d')), '7 days')
  })

  it('can overwrite the unit measures in the initializer', function () {
    var h = humanizer({
      unitMeasures: {
        y: 10512000000,
        mo: 864000000,
        w: 144000000,
        d: 28800000,
        h: 3600000,
        m: 60000,
        s: 1000,
        ms: 1
      }
    })

    assert.equal(h(1000), '1 second')
    assert.equal(h(3600000), '1 hour')
    assert.equal(h(28800000), '1 day')
    assert.equal(h(144000000), '1 week')
  })

  it('can change the decimal', function () {
    var h = humanizer({
      units: ['s'],
      decimal: 'what'
    })

    assert.equal(h(1234), '1what234 seconds')
    assert.equal(h(1234, {
      decimal: '!!'
    }), '1!!234 seconds')
  })

  it('can do simple rounding', function () {
    var h = humanizer({ round: true })

    assert.equal(h(0), '0 seconds')
    assert.equal(h(499), '0 seconds')
    assert.equal(h(500), '1 second')
    assert.equal(h(1000), '1 second')
    assert.equal(h(1499), '1 second')
    assert.equal(h(1500), '2 seconds')
    assert.equal(h(1500), '2 seconds')
    assert.equal(h(121499), '2 minutes, 1 second')
    assert.equal(h(121500), '2 minutes, 2 seconds')
  })

  it('can do rounding with the "units" option', function () {
    var h = humanizer({ round: true })

    assert.equal(h(86364000, { units: ['y', 'mo', 'w', 'd', 'h'] }), '1 day')
    assert.equal(h(1209564000, { units: ['y', 'mo', 'w', 'd', 'h'] }), '2 weeks')
    assert.equal(h(3692131200000, { units: ['y', 'mo'] }), '117 years')
    assert.equal(h(3692131200001, { units: ['y', 'mo', 'w', 'd', 'h', 'm'] }), '116 years, 11 months, 4 weeks, 1 day, 4 hours, 30 minutes')
  })

  it('can do rounding with the "largest" option', function () {
    var h = humanizer({ round: true })

    assert.equal(h(3692131200000, { largest: 1 }), '117 years')
    assert.equal(h(3692131200000, { largest: 2 }), '117 years')
    assert.equal(h(3692131200001, { largest: 100 }), '116 years, 11 months, 4 weeks, 1 day, 4 hours, 30 minutes')
    assert.equal(h(2838550, { largest: 3 }), '47 minutes, 19 seconds')
  })

  it('can ask for the largest units', function () {
    var h = humanizer({ largest: 2 })

    assert.equal(h(0), '0 seconds')
    assert.equal(h(1000), '1 second')
    assert.equal(h(2000), '2 seconds')
    assert.equal(h(540360012), '6 days, 6 hours')
    assert.equal(h(540360012), '6 days, 6 hours')
    assert.equal(h(540360012, { largest: 3 }), '6 days, 6 hours, 6 minutes')
    assert.equal(h(540360012, { largest: 100 }), '6 days, 6 hours, 6 minutes, 0.012 seconds')
  })

  it('has properties which can be modified', function () {
    var h = humanizer()

    assert.equal(h(363000), '6 minutes, 3 seconds')

    h.delimiter = '+'
    assert.equal(h(363000), '6 minutes+3 seconds')

    h.language = 'ur'
    assert.equal(h(363000), '6 منٹ+3 سیکنڈ')

    h.units = ['m']
    assert.equal(h(363000), '6.05 منٹ')
  })

  it('is a named function', function () {
    assert.equal(humanizer().name, 'humanizer')
  })

  it('can add a new language', function () {
    var h = humanizer({ language: 'cool language' })
    h.languages['cool language'] = {
      y: function () { return 'y' },
      mo: function () { return 'mo' },
      w: function () { return 'w' },
      d: function () { return 'd' },
      h: function () { return 'h' },
      m: function () { return 'mi' },
      s: function () { return 's' },
      ms: function () { return 'ms' }
    }

    assert.equal(h(1000), '1 s')
    assert.equal(h(1000, {
      language: 'ur'
    }), '1 سیکنڈ')

    var anotherH = humanizer({
      language: 'cool language'
    })
    assert.throws(function () {
      anotherH(1000)
    }, Error)
  })

  it('can overwrite an existing language', function () {
    var h = humanizer({
      language: 'en'
    })

    assert.equal(h(1000), '1 second')

    h.languages['en'] = {
      y: function () { return 'y' },
      mo: function () { return 'mo' },
      w: function () { return 'w' },
      d: function () { return 'd' },
      h: function () { return 'h' },
      m: function () { return 'm' },
      s: function () { return 's' },
      ms: function () { return 'ms' }
    }

    assert.equal(h(1000), '1 s')
    assert.equal(h(15600000), '4 h, 20 m')

    var anotherH = humanizer({ language: 'en' })

    assert.equal(anotherH(1000), '1 second')
  })

  it('can overwrite the languages property in the initializer', function () {
    var h = humanizer({
      languages: {
        en: {
          y: function () { return 'y' },
          mo: function () { return 'mo' },
          w: function () { return 'w' },
          d: function () { return 'd' },
          h: function () { return 'h' },
          m: function () { return 'm' },
          s: function () { return 's' },
          ms: function () { return 'ms' }
        }
      }
    })

    assert.equal(h(1000), '1 s')
    assert.equal(h(15600000), '4 h, 20 m')
    assert.equal(h(1000, { language: 'ur' }), '1 سیکنڈ')
  })
})


humanize-duration.js
// HumanizeDuration.js - https://git.io/j0HgmQ

;(function () {
  var languages = {
    ar: {
      y: function (c) { return c === 1 ? 'سنة' : 'سنوات' },
      mo: function (c) { return c === 1 ? 'شهر' : 'أشهر' },
      w: function (c) { return c === 1 ? 'أسبوع' : 'أسابيع' },
      d: function (c) { return c === 1 ? 'يوم' : 'أيام' },
      h: function (c) { return c === 1 ? 'ساعة' : 'ساعات' },
      m: function (c) { return c === 1 ? 'دقيقة' : 'دقائق' },
      s: function (c) { return c === 1 ? 'ثانية' : 'ثواني' },
      ms: function (c) { return c === 1 ? 'جزء من الثانية' : 'أجزاء من الثانية' },
      decimal: ','
    },
    en: {
      y: function (c) { return 'year' + (c === 1 ? '' : 's') },
      mo: function (c) { return 'month' + (c === 1 ? '' : 's') },
      w: function (c) { return 'week' + (c === 1 ? '' : 's') },
      d: function (c) { return 'day' + (c === 1 ? '' : 's') },
      h: function (c) { return 'hour' + (c === 1 ? '' : 's') },
      m: function (c) { return 'minute' + (c === 1 ? '' : 's') },
      s: function (c) { return 'second' + (c === 1 ? '' : 's') },
      ms: function (c) { return 'millisecond' + (c === 1 ? '' : 's') },
      decimal: '.'
    },
    ur: {
      y: 'سال',
      mo: function (c) { return c === 1 ? 'مہینہ' : 'مہینے' },
      w: function (c) { return c === 1 ? 'ہفتہ' : 'ہفتے' },
      d: 'دن',
      h: function (c) { return c === 1 ? 'گھنٹہ' : 'گھنٹے' },
      m: 'منٹ',
      s: 'سیکنڈ',
      ms: 'ملی سیکنڈ',
      decimal: '.'
    }
  }

  // You can create a humanizer, which returns a function with default
  // parameters.
  function humanizer (passedOptions) {
    var result = function humanizer (ms, humanizerOptions) {
      var options = extend({}, result, humanizerOptions || {})
      return doHumanization(ms, options)
    }

    return extend(result, {
      language: 'en',
      delimiter: ', ',
      spacer: ' ',
      conjunction: '',
      serialComma: true,
      units: ['y', 'mo', 'w', 'd', 'h', 'm', 's'],
      languages: {},
      round: false,
      unitMeasures: {
        y: 31557600000,
        mo: 2629800000,
        w: 604800000,
        d: 86400000,
        h: 3600000,
        m: 60000,
        s: 1000,
        ms: 1
      }
    }, passedOptions)
  }

  // The main function is just a wrapper around a default humanizer.
  var humanizeDuration = humanizer({})

  // doHumanization does the bulk of the work.
  function doHumanization (ms, options) {
    var i, len, piece

    // Make sure we have a positive number.
    // Has the nice sideffect of turning Number objects into primitives.
    ms = Math.abs(ms)

    var dictionary = options.languages[options.language] || languages[options.language]
    if (!dictionary) {
      throw new Error('No language ' + dictionary + '.')
    }

    var pieces = []

    // Start at the top and keep removing units, bit by bit.
    var unitName, unitMS, unitCount
    for (i = 0, len = options.units.length; i < len; i++) {
      unitName = options.units[i]
      unitMS = options.unitMeasures[unitName]

      // What's the number of full units we can fit?
      if (i + 1 === len) {
        unitCount = ms / unitMS
      } else {
        unitCount = Math.floor(ms / unitMS)
      }

      // Add the string.
      pieces.push({
        unitCount: unitCount,
        unitName: unitName
      })

      // Remove what we just figured out.
      ms -= unitCount * unitMS
    }

    var firstOccupiedUnitIndex = 0
    for (i = 0; i < pieces.length; i++) {
      if (pieces[i].unitCount) {
        firstOccupiedUnitIndex = i
        break
      }
    }

    if (options.round) {
      var ratioToLargerUnit, previousPiece
      for (i = pieces.length - 1; i >= 0; i--) {
        piece = pieces[i]
        piece.unitCount = Math.round(piece.unitCount)

        if (i === 0) { break }

        previousPiece = pieces[i - 1]

        ratioToLargerUnit = options.unitMeasures[previousPiece.unitName] / options.unitMeasures[piece.unitName]
        if ((piece.unitCount % ratioToLargerUnit) === 0 || (options.largest && ((options.largest - 1) < (i - firstOccupiedUnitIndex)))) {
          previousPiece.unitCount += piece.unitCount / ratioToLargerUnit
          piece.unitCount = 0
        }
      }
    }

    var result = []
    for (i = 0, pieces.length; i < len; i++) {
      piece = pieces[i]
      if (piece.unitCount) {
        result.push(render(piece.unitCount, piece.unitName, dictionary, options))
      }

      if (result.length === options.largest) { break }
    }

    if (result.length) {
      if (!options.conjunction || result.length === 1) {
        return result.join(options.delimiter)
      } else if (result.length === 2) {
        return result.join(options.conjunction)
      } else if (result.length > 2) {
        return result.slice(0, -1).join(options.delimiter) + (options.serialComma ? ',' : '') + options.conjunction + result.slice(-1)
      }
    } else {
      return render(0, options.units[options.units.length - 1], dictionary, options)
    }
  }

  function render (count, type, dictionary, options) {
    var decimal
    if (options.decimal === void 0) {
      decimal = dictionary.decimal
    } else {
      decimal = options.decimal
    }

    var countStr = count.toString().replace('.', decimal)

    var dictionaryValue = dictionary[type]
    var word
    if (typeof dictionaryValue === 'function') {
      word = dictionaryValue(count)
    } else {
      word = dictionaryValue
    }

    return countStr + options.spacer + word
  }

  function extend (destination) {
    var source
    for (var i = 1; i < arguments.length; i++) {
      source = arguments[i]
      for (var prop in source) {
        if (source.hasOwnProperty(prop)) {
          destination[prop] = source[prop]
        }
      }
    }
    return destination
  }

  humanizeDuration.getSupportedLanguages = function getSupportedLanguages () {
    var result = []
    for (var language in languages) {
      if (languages.hasOwnProperty(language)) {
        result.push(language)
      }
    }
    return result
  }

  humanizeDuration.humanizer = humanizer

  if (typeof define === 'function' && define.amd) {
    define(function () {
      return humanizeDuration
    })
  } else if (typeof module !== 'undefined' && module.exports) {
    module.exports = humanizeDuration
  } else {
    this.humanizeDuration = humanizeDuration
  }
})(); // eslint-disable-line semi
